src/App.tsx

```diff
@@ -15,36 +18,122 @@
+  }
+
+  return preparedGoods;
+};
+
 export const App: React.FC = () => {
+  const [filterInstruction, setFilterInstruction] = React.useState({
+    filterType: FilterTypes.default,
+    isReversed: false,
+  });
+
+  const isAnyFilterApplied =
+    filterInstruction.filterType || filterInstruction.isReversed;
+
+  const visibleGoods = getPreparedGoods(goodsFromServer, filterInstruction);
+
+  const handleSetFilterType = (filterTypeName: FilterTypes) => {
+    setFilterInstruction(prevState => {
+      return {
+        ...prevState,
+        filterType: filterTypeName,
+      };
+    });
+  };
+
+  const handleReverse = () => {
+    setFilterInstruction(prevState => {
+      return {
+        ...prevState,
+        isReversed: !prevState.isReversed,
+      };
+    });
+  };
+
+  const handleClearFilter = () => {
+    setFilterInstruction(prevState => {
+      return {
+        ...prevState,
+        filterType: FilterTypes.default,
+        isReversed: false,
+      };
+    });
+  };
+
+  const isLightOn = (filterType: FilterTypes) => {
+    return filterInstruction.filterType !== filterType;
+  };
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          onClick={() => handleSetFilterType(FilterTypes.byAlphabet)}
+          type="button"
+          className={cn('button', 'is-info', {
+            'is-light': isLightOn(FilterTypes.byAlphabet),
+          })}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          onClick={() => handleSetFilterType(FilterTypes.byLength)}
+          type="button"
+          className={cn('button', 'is-success', {
+            'is-light': isLightOn(FilterTypes.byLength),
+          })}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          onClick={handleReverse}
+          type="button"
+          className={cn('button', 'is-warning', {
+            'is-light': !filterInstruction.isReversed,
+          })}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {isAnyFilterApplied && (
+          <button
+            onClick={() => handleClearFilter()}
+            type="button"
+            className="button is-danger is-light"
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {visibleGoods.map(good => (
+          <li data-cy="Good" key={good}>
```

Comment:
When rendering a list, ensure that the `key` prop is a unique identifier for each element. Using `good` as a key might not be unique if the list contains duplicate items. Consider using an index or another unique piece of data. Refer to the [CODE KNOWLEDGE] checklist.